File: android/src/main/java/com/otasdk/CrashGuard.kt
package com.otasdk

import android.util.Log

/**
 * CrashGuard detects if a newly applied bundle is causing repeated crashes
 * and triggers a rollback to the previous stable bundle.
 *
 * Strategy:
 *  - On each successful app launch, reset the crash counter.
 *  - If the app crashes before reaching "stable" (within first 5 seconds),
 *    the crash counter is NOT reset — increment on next launch.
 *  - If crash count >= threshold, auto-rollback.
 *
 * Usage:
 *  Call CrashGuard.onAppStart() in your Application.onCreate()
 *  Call CrashGuard.markStable() after your app is fully initialised (5+ seconds)
 */
class CrashGuard(
    private val prefs: OTAPrefs,
    private val threshold: Int = 3,
) {
    companion object {
        private const val TAG = "OTA-CrashGuard"
    }

    /**
     * Call this at app start.
     * Returns true if a rollback should be triggered.
     */
    fun onAppStart(): Boolean {
        val pendingHash = prefs.pendingBundlePath
        val lastApplied = prefs.lastAppliedBundleHash
        val currentActive = prefs.activeBundleHash

        // If the active bundle changed since last crash check, reset counter
        if (currentActive != lastApplied) {
            Log.d(TAG, "New bundle detected ($currentActive), resetting crash counter")
            prefs.lastAppliedBundleHash = currentActive
            prefs.resetCrashCount()
            return false
        }

        // Increment crash count (will be reset if markStable() is called)
        val count = prefs.crashCount + 1
        prefs.crashCount = count
        Log.w(TAG, "Crash count for bundle $currentActive: $count / $threshold")

        if (count >= threshold) {
            Log.e(TAG, "Crash threshold reached — triggering rollback")
            return true
        }
        return false
    }

    /**
     * Call this when the app has been running for at least 5 seconds without crash.
     * Resets the crash counter, marking this bundle as stable.
     */
    fun markStable() {
        if (prefs.crashCount > 0) {
            Log.d(TAG, "App stable — resetting crash counter")
            prefs.resetCrashCount()
        }
    }
}

File: ios/CrashGuard.swift
import Foundation

/// Detects crash loops on newly applied bundles and triggers rollback.
class CrashGuard {

    private let prefs: OTAPrefs
    private let threshold: Int

    init(prefs: OTAPrefs, threshold: Int = 3) {
        self.prefs = prefs
        self.threshold = threshold
    }

    /// Call at every app start. Returns true if rollback should happen.
    func onAppStart() -> Bool {
        let currentHash = prefs.activeBundleHash
        let lastHash    = prefs.lastAppliedBundleHash

        // New bundle detected — reset counter
        if currentHash != lastHash {
            NSLog("[OTA-CrashGuard] New bundle (%@), resetting counter", currentHash)
            prefs.lastAppliedBundleHash = currentHash
            prefs.resetCrashCount()
            return false
        }

        // Same bundle — increment crash counter
        let count = prefs.crashCount + 1
        prefs.crashCount = count
        NSLog("[OTA-CrashGuard] Crash count for %@: %d / %d", currentHash, count, threshold)

        if count >= threshold {
            NSLog("[OTA-CrashGuard] Threshold reached — rollback triggered")
            return true
        }
        return false
    }

    /// Call once the app has been running stably (≥5 seconds).
    func markStable() {
        if prefs.crashCount > 0 {
            NSLog("[OTA-CrashGuard] App stable — reset crash counter")
            prefs.resetCrashCount()
        }
    }
}
